des/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+    let b2 = (block & 0x0000000000480000) << 13;
+    let b3 = (block & 0x0000088000000000) << 12;
+    let b4 = (block & 0x0000002020120000) << 25;
+    let b5 = (block & 0x0000000442000000) << 14;
+    let b6 = (block & 0x0000000001800000) << 37;
+    let b7 = (block & 0x0000000004000000) << 24;
+    let b8 = (block & 0x0000020280015000).wrapping_mul(0x0000020080800083) & 0x02000a6400000000;
+    let b9 = (block.rotate_left(29) & 0x01001400000000aa).wrapping_mul(0x0000210210008081) & 0x0902c01200000000;
+    let b10 = (block & 0x0000000910040000).wrapping_mul(0x0000000c04000020) & 0x8410010000000000;
+    b1 | b2 | b3 | b4 | b5 | b6 | b7 | b8 | b9 | b10
+}
+
+/// Generate the 16 subkeys
+fn gen_keys(key: u64) -> [u64; 16] {
+    let mut keys: [u64; 16] = [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
+    let key = pc1(key);
+
+    // The most significant bit is bit zero, and there are only 56 bits in
+    // the key after applying PC1, so we need to remove the eight least
+    // significant bits from the key.
+    let key = key >> 8;
+
+    let mut c = key >> 28;
+    let mut d = key & 0x0FFFFFFF;
+    for i in 0..16 {
+        c = rotate(c, SHIFTS[i]);
+        d = rotate(d, SHIFTS[i]);
+
+        // We need the `<< 8` because the most significant bit is bit zero,
+        // so we need to shift our 56 bit value 8 bits to the left.
+        keys[i] = pc2(((c << 28) | d) << 8 );
+    }
+
+    keys
+}
+
+/// Performs a left rotate on a 28 bit number
+fn rotate(mut val: u64, shift: u8) -> u64 {
+    let top_bits = val >> (28 - shift);
+    val <<= shift;
+
+    (val | top_bits) & 0x0FFFFFFF
+}
+
+fn round(input: u64, key: u64) -> u64 {
+    let l = (input >> 32) << 32;
+    let r = (input & 0xFFFFFFFF) << 32;
+
+    r | ((f(r, key) ^ l) >> 32)
+}
+
+fn f(input: u64, key: u64) -> u64 {
+    let mut val = e(input as u64);
+    val ^= key;
+    val = apply_sboxes(val);
+    p(val)
+}
+
+/// Applies all eight sboxes to the input
+fn apply_sboxes(input: u64) -> u64 {
+    let mut output: u64 = 0;
+    for i in 0..8 {
+        let sbox = SBOXES[i];
+        let val = (input >> (58 - (i * 6))) & 0x3F;
+        output |= (sbox[val as usize] as u64) << (60 - (i * 4));
+    }
+    output
+}
+
+impl Des {
+    fn encrypt(&self, input: u64) -> u64 {
+        let mut data = ip(input);
+        for key in self.keys.iter() {
+            data = round(data, *key);
+        }
+        fp((data & 0x0FFFFFFF) << 32 | (data >> 32))
+    }
+
+    fn decrypt(&self, input: u64) -> u64 {
+        let mut data = ip(input);
+        for key in self.keys.iter().rev() {
+            data = round(data, *key);
+        }
+        fp((data & 0x0FFFFFFF) << 32 | (data >> 32))
+    }
+}
+
+impl BlockCipher for Des {
+    type BlockSize = U8;
+
+    fn encrypt_block(&self, input: &Block<U8>, output: &mut Block<U8>) {
+        let block = read_u64_be(input);
+        let res = self.encrypt(block);
+        write_u64_be(output, res);
```

Comment:
I've added `read_u64_be` to `byte-tools`. Don't forget to run `cargo update` to get new version.